[Clang] Normalize constraints before checking for satisfaction

cor3ntin opened the pull request:
In the standard, constraint satisfaction checking is done on the normalized form of a constraint.

Clang instead substitutes on the non-normalized form, which causes us to report substitution failures in template arguments or concept ids, which is non-conforming but unavoidable without a parameter mapping

This patch normalizes before satisfaction checking. However, we preserve concept-id nodes in the normalized form, solely for diagnostics purposes.

This addresses #61811 and related concepts conformance bugs, ideally to make the remaining implementation of concept template parameters easier

Fixes #135190
Fixes  #61811

Co-authored-by: Younan Zhang <zyn7109@gmail.com>

shafik commented on clang/include/clang/Sema/SemaConcept.h:
```diff
@@ -16,130 +16,362 @@
-struct alignas(ConstraintAlignment) FoldExpandedConstraint;
+public:
+  ConstraintKind getKind() const {
+    return static_cast<ConstraintKind>(Atomic.Kind);
```
Maybe worth commenting this is only valid due to initial common sequence.

shafik commented on clang/include/clang/Sema/SemaConcept.h:
```diff
@@ -16,130 +16,362 @@
+  }
+
+  bool hasParameterMapping() const {
+    assert(getKind() != ConstraintKind::Compound);
```
I am curious why you only assert against `ConstraintLind::Compound`

shafik reviewed (commented):
I need to make another pass later.

cor3ntin commented:
There are still a few test failures, including in libc++, but I think we are are at the stage where this could benefit from more eyes

cor3ntin committed:
Use normalization for concept satisfaction

cor3ntin committed:
more build failures

cor3ntin committed:
I accidentally disabled caching

cor3ntin committed:
fix libcxx test

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -543,30 +672,330 @@ static ExprResult calculateConstraintSatisfaction(
+
+  ExprResult Out;
+  UnsignedOrNone NumExpansions =
+      EvaluateFoldExpandedConstraintSize(Constraint, *SubstitutedArgs);
+  if (!NumExpansions)
+    return ExprEmpty();
```
This includes a parameter mapping substitution which is expensive, so let's add a TODO of adding cache on it.

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -543,30 +672,330 @@ static ExprResult calculateConstraintSatisfaction(
+
+  llvm::SmallVector<TemplateArgument> SubstitutedOuterMost;
+  std::optional<MultiLevelTemplateArgumentList> SubstitutedArgs =
+      SubstitutionInTemplateArguments(Constraint, MLTAL, SubstitutedOuterMost);
+
+  if (!SubstitutedArgs) {
+    Satisfaction.IsSatisfied = false;
+    // FIXME: diagnostics?
+    return ExprError();
+  }
```
This is also expensive, caching it might benefit the SFINAE path.

cor3ntin replied:
I added a fixme above

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -543,30 +672,330 @@ static ExprResult calculateConstraintSatisfaction(
+  SS.Adopt(Constraint.getConceptId()->getNestedNameSpecifierLoc());
 
-  return SubstitutedAtomicExpr;
+  ExprResult SubstitutedConceptId = S.CheckConceptTemplateId(
+      SS, Constraint.getConceptId()->getTemplateKWLoc(),
```
Maybe we don't need to rebuild a ConceptSpecializationDecl - again I forgot to clean it up

cor3ntin replied:
We do, because we need to have the right expression to display in diagnostics

zyn0217 commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -51,6 +51,27 @@ class FixedPointSemantics;
 struct fltSemantics;
 template <typename T, unsigned N> class SmallPtrSet;
 
+template <> struct DenseMapInfo<llvm::FoldingSetNodeID> {
```
Maybe split it into an NFC PR? I'm not sure if downstream users want it

zyn0217 commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -352,6 +356,8 @@ class TreeTransform {
   /// being expanded.
   void ExpandingFunctionParameterPack(ParmVarDecl *Pack) { }
 
+  bool ShouldPreserveTemplateArgumentsPacks() const { return false; }
+
```
This is no longer needed afaik :)

zyn0217 reviewed (commented):
Thank you very much for putting it up and the incredible work!

I left some notes, mostly are those that I didn't get around to yet (and I think they should be useful); so maybe we want to resolve them before we move forward.

Also let's not forget the release note :)

cor3ntin committed:
fix windows build

cor3ntin committed:
format

cor3ntin commented:
Update: still needs cleaning but the test pass

cor3ntin committed:
fix libc++ test

cor3ntin commented on clang/include/clang/AST/ASTConcept.h:
```diff
@@ -27,13 +27,23 @@
+/// substituted constraint expr, if the template arguments could be
+/// substituted into them, or a diagnostic if substitution resulted in
+/// an invalid expression.
+using UnsatisfiedConstraintRecord =
```
The name is pre-existing, we could change it afterwards

cor3ntin committed:
cleanups

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -543,30 +672,330 @@ static ExprResult calculateConstraintSatisfaction(
+    Satisfaction.IsSatisfied = Cached.IsSatisfied;
+    Satisfaction.Details.insert(Satisfaction.Details.begin() + Size,
+                                Cached.Details.begin(), Cached.Details.end());
+    return Iter->second.SubstExpr;
+  }
```
I don;t think we want to upstream that, it was just too slow to be practical, even as a debug tool

cor3ntin committed:
add fixmes

cor3ntin committed:
fix windows build

zyn0217 reviewed (approved):
A few more comments, and it would be great if we could tidy up our header includes.

Please wait for some time before others want to speak something. Otherwise this LGTM assuming CI is happy with our approach. 😄

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -543,30 +661,328 @@ static ExprResult calculateConstraintSatisfaction(
+    PartialDiagnosticAt SubstDiag{SourceLocation(),
+                                  PartialDiagnostic::NullDiagnostic()};
+    Info.takeSFINAEDiagnostic(SubstDiag);
+    // FIXME: Concepts: This is an unfortunate consequence of there
+    //  being no serialization code for PartialDiagnostics and the fact
+    //  that serializing them would likely take a lot more storage than
+    //  just storing them as strings. We would still like, in the
+    //  future, to serialize the proper PartialDiagnostic as serializing
+    //  it as a string defeats the purpose of the diagnostic mechanism.
```
Ideally we would never store diags in the ast that way!

cor3ntin committed:
more cleanups

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -543,30 +661,328 @@ static ExprResult calculateConstraintSatisfaction(
+  }
+
+  CXXScopeSpec SS;
+  SS.Adopt(Constraint.getConceptId()->getNestedNameSpecifierLoc());
```
Looking at it again I feel that we are repeating `Constraint.getConceptId()`.

Can we have a handy variable of it at the start of the function?

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -10,19 +10,68 @@
+ * template <typename U>
+ * concept B = A<U> && __is_same(U, int);
+ *
+ * The  normal form is `__is_same(T, int) /T->U, inner most level/
+ *                   && __is_same(U, int) {U->U} /T->U, outermost most level/
+ *                   `
```
The normal form of B<U> is ...

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -10,19 +10,68 @@
 //
 //===----------------------------------------------------------------------===//
 
+/*
+ * A note on implementation:
+ *
```
thanks, the documentation is great

zyn0217 commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -822,14 +866,17 @@ ExprResult CalculateConstraintSatisfaction::Calculate(
 
+  // ConceptIdConstraint is only relevant for diagnostics,
+  // so if the normalized constraint is satisfied, we should not
+  // substiture into the constraint.
```
```suggestion
  // substitute into the constraint.
```

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1405,7 +1405,8 @@ class TemplateInstantiator : public TreeTransform<TemplateInstantiator> {
 
-  bool PreserveArgumentPacks = false;
+  // Whether to rebuild pack expansion types; We don't do that when
+  // rebuilding a fold expression appearing in a constraint expression.
```
when building parameter mappings for a fold expression appearing in a constraint expression.

(Let's be specific)

cor3ntin committed:
remove some whitespace changes

cor3ntin committed:
fix doc

cor3ntin committed:
more cleanups

cor3ntin committed:
remove some duplication

cor3ntin committed:
Cache satisfaction of ConceptIdConstraint.

We substitute into concept ids for diagnostics purposes.
However, most of the time this happen in SFINEA context
and impact performance noticeably.

cor3ntin committed:
more caching

zyn0217 committed:
Fix stdexec regressions (#59)

We may substitute _Fn::template __f (a dependent template name) into
a template-template parameter name, when building a parameter mapping.

They can't have the same NNSLoc, so the assertion doesn't hold.

Also we missed a case when profiling the concept, as in
InjectedClassNameType.

cor3ntin committed:
more cleanups

cor3ntin committed:
Release note

mizvekov reviewed (commented):
Just a first pass, I'll try to get some time to fully review later.

zyn0217 commented on clang/test/SemaTemplate/concepts.cpp:
```diff
@@ -1257,4 +1256,35 @@ template <typename T> concept PerfectSquare = [](){} // expected-note 2{{here}}
+
+namespace parameter_mapping_regressions {
+
+namespace case1 {
+
+template <template <class> class> using __meval = struct __q;
+template <template <class> class _Tp>
+concept __mvalid = requires { typename __meval<_Tp>; };
+template <class _Fn>
+concept __minvocable = __mvalid<_Fn::template __f>;
+template <class...> struct __mdefer_;
+template <class _Fn, class... _Args>
+  requires __minvocable<_Fn>
+struct __mdefer_<_Fn, _Args...> {};
+template <class = __q> struct __mtransform {
+  template <class> using __f = int;
+};
+struct __completion_domain_or_none_ : __mdefer_<__mtransform<>> {};
+
+}
```
@mizvekov Here is the regression why I removed that assert in TemplateArgumentLoc.

So basically we want to build the parameter mapping of __minvocable<Fn> from line 1281. We use template rewrite mechanism, and we end up substituting `_Fn::template __f` into template template parameter on line 1275.

I noticed your last removal of DTST patch and I tried with it locally, but unfortunately it doesn't fix that assertion failure where that template template parameter doesn't have the same NNS with the rewritten template argument.

The rewritten template argument comes from CheckTemplateArguments, which is necessary to correctly preserve packs in other cases.

We chose to use rewritten mode in very early developing stage because we had trouble with Subst nodes and lambdas, and it was impossible to fix that without rewriting half of clang.

I tried rebuilding that TemplateName with its NNS cleared, but it turned out I have to handle so many types in Instantiator which is a bit clumsy. (And I failed some regressions too!)

Maybe we are having the same problem with CTAD guides, but I didn't come up with a case so far.

Do you have any suggestions for such a case?

mizvekov replied:
Moving the discussion to the offending line.

zyn0217 commented on clang/lib/AST/TemplateBase.cpp:
```diff
@@ -606,8 +606,11 @@ TemplateArgumentLoc::TemplateArgumentLoc(ASTContext &Ctx,
          Argument.getKind() == TemplateArgument::TemplateExpansion);
-  assert(QualifierLoc.getNestedNameSpecifier() ==
-         Argument.getAsTemplateOrTemplatePattern().getQualifier());
+  // We can't assume QualifierLoc.getNestedNameSpecifier() ==
+  // Argument.getAsTemplateOrTemplatePattern().getQualifier() at this point,
+  // because in template rewriting, we may substitute a DependentTemplateName
+  // (which has a NNSLoc) into a template template parameter (which
+  // doesn't have a NNSLoc).
```
I explained why under the corresponding regression case. Let's move our conversation there

zyn0217 commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -126,6 +126,10 @@ C++23 Feature Support
 C++20 Feature Support
 ^^^^^^^^^^^^^^^^^^^^^
 
+- Clang now normalizes constraints before checking their satisfaction, as mandated by the standard.
+  As a result, Clang no longer incorrectly diagnose subsdtitution failures in template arguments only
+  used in concept-ids, and produce better diagnostics for satisfaction failure. (#G61811) (#GH135190)
```
```suggestion
- Clang now normalizes constraints before checking their satisfaction, as mandated by the standard.
  As a result, Clang no longer incorrectly diagnose subsdtitution failures in template arguments only
  used in concept-ids, and produce better diagnostics for satisfaction failure. (#GH61811) (#GH135190)
```

mizvekov commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1379,216 +1890,327 @@ static void diagnoseWellFormedUnsatisfiedConstraintExpr(Sema &S,
+        /*Pattern=*/nullptr,
+        /*ForConstraintInstantiation=*/true);
+
+    // Don't build Subst* nodes to model lambda expressions.
+    // The transform of Subst* is oblivious to the lambda type.
+    MLTAL.setKind(TemplateSubstitutionKind::Rewrite);
```
@zyn0217 moving the discussion here.

I think this is incorrect, this substitution is not a rewrite, as this can basically transform a reference to template parameter into an arbitrary template argument of the corresponding kind.

A rewrite should transform a template parameter into another template parameter of the same kind, otherwise the source location kinds will be incompatible.

For example, in a non-rewrite substitution, the source location for a template parameter is compatible with the source location for its corresponding Subst* AST node kind. This makes sense, as the Subst node is a stand-in for the template parameter, and whatever it was substituted for is not part of the source location data, as it was not written.

For a rewrite, if we stick to the rule of substituting template parameters for template parameters, the source locations obviously stay compatible as they are of the same kind.
But if you allow arbitrary substitutions outside of those rules, then that doesn't work anymore.

We should add some asserts to verify this isn't happening anywhere else and doesn't happen again.

If you add back the assert in the TemplateArgumentLoc constructor, and remove this line, that should make it work.

In fact, the otherwise failing test case can be simplified to just:
```C++
template <template <class> class TT> concept C1 = sizeof(TT<int>) >= 0;
template <class T> concept C2 = C1<T::template X>;
static_assert(C2<int>);
```

I am not sure why this line is needed in the first place, that looks like a workaround.

zyn0217 replied:
We avoided Subst* nodes because we encountered the following case:

```cpp
template<auto Q> concept C = requires { Q.template operator()<float>(); };
template<class> concept D = false;
template<class> concept E = C<[]<D Ty>{}>;
static_assert(E<int>);
```

For Q, we build a SubstNTTPExpr with a dependent lambda `[]<D Ty>{}`. And since we still lack the parameter mapping inside a requirement expression (https://eel.is/c++draft/temp.constr.normal#1.6) so the lambda will remain dependent after we substitute <int> into the mapping, and therefore evaluating it would be UB.

I am still not very clear if we need a parameter mapping inside C. @cor3ntin

zyn0217 replied:
Actually it can be even simplified to

```cpp
template<auto Q> concept C = Q.template operator()<float>();
template<class> concept E = C<[]<class Ty>{ return false; }>;
static_assert(E<int>);
```

And we cannot perform a qualified name lookup on a subst* node.

zyn0217 committed:
Use ordinary substitution for parameter mapping (#60)

cor3ntin committed:
fix typos

cor3ntin committed:
fix tests after rebase

cor3ntin commented on clang/lib/AST/ASTConcept.cpp:
```diff
@@ -116,6 +122,19 @@ void ConceptReference::print(llvm::raw_ostream &OS,
+                                             const ConceptReference *C) {
+  std::string NameStr;
+  llvm::raw_string_ostream OS(NameStr);
+  LangOptions LO;
```
Yeah, we do that in a couple of places to avoid that dependence - I agree about it being gnarly.

cor3ntin committed:
address some of Erich's feedback

cor3ntin committed:
fix rebase

cor3ntin committed:
remove comment (moved to doc)

cor3ntin commented on clang/include/clang/Sema/SemaConcept.h:
```diff
@@ -16,130 +16,385 @@
-  std::optional<ArrayRef<TemplateArgumentLoc>> ParameterMapping;
+  const OccurenceList &mappingOccurenceList() const {
+    assert(hasParameterMapping() && "This constraint has no parameter mapping");
+    return Atomic.Indexes;
```
It's copied entirely in in `SubstituteParameterMappings::substitute`, so it needs to be exposed

zyn0217 commented on clang/docs/InternalsManual.rst:
```diff
@@ -2859,6 +2859,65 @@ This library is called by the :ref:`Parser library <Parser>` during parsing to
+contains a parameter mapping, represented by the indexes of all parameter
+being used.
+
+Checking satisfaction is done by ConstraintSatisfactionChecker, recursively
+walking NormalizedConstraint. At each level, we substitute the outermost
```
```suggestion
Checking satisfaction is done by ``ConstraintSatisfactionChecker``, recursively
walking ``NormalizedConstraint``. At each level, we substitute the outermost
```

zyn0217 commented on clang/docs/InternalsManual.rst:
```diff
@@ -2859,6 +2859,65 @@ This library is called by the :ref:`Parser library <Parser>` during parsing to
+Checking satisfaction is done by ConstraintSatisfactionChecker, recursively
+walking NormalizedConstraint. At each level, we substitute the outermost
+level of the template arguments referenced in the parameter mapping of a
+normalized expression (``MultiLevelTemplateArgumentList``).
+
+.. code-block:: c++
```
```suggestion
normalized expression (``MultiLevelTemplateArgumentList``).

For the following example,

.. code-block:: c++
```

zyn0217 commented on clang/docs/InternalsManual.rst:
```diff
@@ -2859,6 +2859,65 @@ This library is called by the :ref:`Parser library <Parser>` during parsing to
+ && __is_same(U, int) {U->U} /*T->U, outermost most level*/
+
+After substitution in the mapping, we substitute in the constraint expression
+using that copy of the MultiLevelTemplateArgumentList, and then evaluate it.
```
```suggestion
using that copy of the ``MultiLevelTemplateArgumentList``, and then evaluate it.
```

zyn0217 reviewed (commented):
I added some comments for document formats

cor3ntin committed:
Make sure the SubstitutionDiagnostic alias is used consistently

cor3ntin committed:
fix typos, fix merge

cor3ntin commented on clang/include/clang/AST/ASTConcept.h:
```diff
@@ -28,10 +28,19 @@ namespace clang {
+/// an invalid expression.
+using UnsatisfiedConstraintRecord =
+    llvm::PointerUnion<const Expr *, const ConceptReference *,
+                       std::pair<SourceLocation, StringRef> *>;
```
I cleaned that up to hopefully make it more obvious what is going on - it is a union of 3 pointers - the pair is pre-existing though

cor3ntin replied:
None that i would like to get rid of that whole thing entirely and not store diags in the AST but I have not explored how viable this would be

cor3ntin commented on clang/include/clang/Sema/SemaConcept.h:
```diff
@@ -16,130 +16,385 @@
+    CCK_Conjunction,
+    CCK_Disjunction
+  };
+  enum class FoldOperatorKind : unsigned char { And, Or };
```
They all are different things. I would prefer they remain separate to match the wording

A constraint can be either a Compound constraint, or a FoldExpanded constraint.
a Compound constraint can be either a conjunction or a disjunction.

a FoldExpanded has an operator. the reason we have an enum for that instead of reusing `BinaryOperatorKind` is that we don't need/want to use more than one bits (and none of the operator produce a fold expanded constraint)

cor3ntin committed:
fix docs

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1625,6 +1647,13 @@ namespace {
+                                   ElaboratedTypeKeyword Keyword,
+                                   NestedNameSpecifierLoc QualifierLoc,
+                                   QualType T);
+
```
This is not used anywhere

cor3ntin committed:
remove unused code

zyn0217 commented on clang/include/clang/Sema/SemaConcept.h:
```diff
@@ -16,130 +16,385 @@
+
+  using OccurenceList = llvm::SmallBitVector;
+
+  friend bool
+  substituteParameterMappings(Sema &S, NormalizedConstraint &N,
+                              const MultiLevelTemplateArgumentList &MLTAL,
+                              const ASTTemplateArgumentListInfo *ArgsAsWritten,
+                              TemplateParameterList *TemplateParams,
+                              const NamedDecl *D);
```
I think this is unused

cor3ntin committed:
remove more unused code

cor3ntin committed:
fix docs

cor3ntin committed:
format

cor3ntin committed:
Deal with parameters used, but not referenced by an atomic constraint

These parameters should not affect subsumption.
This force us to maintain two occurence list for each atomic constraint.

cor3ntin committed:
oups

cor3ntin committed:
Fix assert, address feedback, add tests

cor3ntin committed:
pack tests, remove dead code

cor3ntin committed:
fix tests for c++20

cor3ntin committed:
restore code that was actually useful

cor3ntin committed:
Merge remote-tracking branch 'llvm_be_very_careful/main' into corentin/use_normalization_for_satisfaction

cor3ntin committed:
Merge remote-tracking branch 'llvm_be_very_careful/main' into corentin/use_normalization_for_satisfaction

cor3ntin committed:
format

cor3ntin committed:
Merge remote-tracking branch 'llvm_be_very_careful/main' into corentin/use_normalization_for_satisfaction

kstoimenov commented:
Looks like it has a memory leak: https://lab.llvm.org/buildbot/#/builders/169/builds/15562. @cor3ntin please take a look.
